### PR TITLE
Add support for models with mutated buffer on torch.onnx.dynamo_export

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114407
* __->__ #112272

This PR adds a unit test that leverages `torch.export.ExportedProgram` models that mutates registered buffers. Although the exporter already works out of the box in such scenario, the GraphModule and the exported ONNX model have extra outputs containing the mutated buffers. On future runs of the ONNX model, the mutated buffers are used as input to the model.

The aforementioned extra inputs and outputs are by design and the `ONNXProgram.model_signature` can be used to fetch detailed input/output schema for the exported model.

However, when we want to compare pytorch output to ONNX's, there is a mismatch between the schema because pytorch output does not include the mutated buffers present on the ONNX output.

This PR extends `onnx_program.adapt_torch_outputs_to_onnx(torch_outputs)` so that the mutated buffers are prepended to the Pytorch output, matching the ONNX schema.

### CODE REVIEW

File: test/onnx/onnx_test_common.py [hunk 1]
```diff
@@ -424,11 +426,22 @@ def _compare_pytorch_onnx_with_ort(
     )
     ort_outputs = run_ort(onnx_program, onnx_format_args)
 
-    if len(ref_outputs) != len(ort_outputs):
+    # When model is a torch.export.ExportedProgram, the number of outputs in the ONNX model can be greater
```
Comment (BowenBao) [inline test/onnx/onnx_test_common.py:429]: Create an output adapter step for this?

Comment (thiagocrepaldi) [reply test/onnx/onnx_test_common.py:429]: It doesn't work, at least how we designed ` InputAdapter`  and `OutputAdapter`. They were designed to be used *after* the model is exported, to adapt input/output from pytorch to match the ONNX input/output.

The output adapter expand the pytorch output to match the ONNX output schema, so we could at most match the number of outputs. Thus, an extra output would be prepended to pytorch output to match the extra output on onnx graph.

However, the ONNX output will have an output with the *mutated* buffer while pytorch will have a not-mutated buffer as output, causing a mismatch between them.

Now if we decide to hack the purpose of the `OutputAdapter` and use it before the FX-ONNX to truncate the extra outputs, we could make it work. This is a break of contract, though.

Makes sense?

Comment (thiagocrepaldi) [reply test/onnx/onnx_test_common.py:429]: I am creating a `adapt_onnx_outputs_to_torch` in a separate PR to help bridging this gap. Despite the output truncation, we will also need to update the mutated buffers within the model so that buffers match during tests

Comment (thiagocrepaldi) [reply test/onnx/onnx_test_common.py]: done

**Comment by BowenBao** [comment]
ci failures?

File: test/onnx/dynamo/test_exporter_api.py [hunk 2]
```diff
@@ -190,6 +190,7 @@ def test_raise_on_invalid_save_argument_type(self):
             onnx.ModelProto(),
             io_adapter.InputAdapter(),
             io_adapter.OutputAdapter(),
+            io_adapter.OutputAdapter(),
```
Comment (BowenBao) [inline test/onnx/dynamo/test_exporter_api.py:193]: this looks weird

Comment (BowenBao) [reply test/onnx/dynamo/test_exporter_api.py:193]: oh i see you are adding another output adapter in this PR.

File: test/onnx/onnx_test_common.py [hunk 3]
```diff
@@ -459,14 +459,24 @@ def _compare_pytorch_onnx_with_ort(
...
-            ref_output, torch.tensor(ort_output), rtol=rtol, atol=atol
-        )
+    for output_format in (
+        "pytorch",
+        "onnx",
+    ):  # Test output match with pytorch and onnx output format
+        if output_format == "pytorch":
+            ort_outputs = onnx_program.adapt_onnx_outputs_to_torch(ort_outputs)
+        elif output_format == "onnx":
+            ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_outputs)
```
Comment (BowenBao) [inline test/onnx/onnx_test_common.py:469]: this is done once above on L451

File: test/onnx/onnx_test_common.py [hunk 4]
```diff
@@ -459,14 +459,24 @@ def _compare_pytorch_onnx_with_ort(
...
-    for ref_output, ort_output in zip(ref_outputs, ort_outputs):
-        torch.testing.assert_close(
-            ref_output, torch.tensor(ort_output), rtol=rtol, atol=atol
-        )
+    for output_format in (
+        "pytorch",
+        "onnx",
+    ):  # Test output match with pytorch and onnx output format
+        if output_format == "pytorch":
+            ort_outputs = onnx_program.adapt_onnx_outputs_to_torch(ort_outputs)
```
Comment (BowenBao) [inline test/onnx/onnx_test_common.py:467]: I didn't do this because it felt non-trivial. onnx outputs are more organized. It can only be tensors, or (nested) list of tensors. Whereas for torch, the output can be anything. So mapping back from onnx to torch is considerably harder (how to insert back the Nones at the correct location where we discarded? how to construct back the arbitrary torch model output classes/datastructures that we flattened?).

I see that the goal was rather to eliminate the extra mutated buffers. I feel finding the way to include those outputs, when they exist because we are exporting ExportProgram, in the `ref_outputs` and have them properly accuracy checked might be more desirable.

File: torch/onnx/_internal/io_adapter.py [hunk 5]
```diff
@@ -507,6 +515,46 @@ def apply(self, model_outputs: Any) -> Sequence[Any]:
...
+            updated_outputs: The model outputs without mutated buffers and user inputs.
+        """
+
+        assert len(model_outputs) == len(
+            self.model.graph_signature.output_specs
+        ), "Actual model outputs mismatches the Model output spec"
+
+        user_outputs = []
+        for out, out_spec in zip(
+            model_outputs, self.model.graph_signature.output_specs
```
Comment (BowenBao) [inline torch/onnx/_internal/io_adapter.py:551]: Similar concern as https://github.com/pytorch/pytorch/pull/113477#discussion_r1394972191

File: test/onnx/dynamo/test_exporter_api.py (hunk 2 above)
Comment (thiagocrepaldi) [reply test/onnx/dynamo/test_exporter_api.py:193]: yes, we need one output adapter to go from torch to onnx format and another to go from onnx to torch.
e.g. onnx->torch we drop extra outputs on the onnx model; torch->onnx we prepend outputs into the torch output

File: torch/onnx/_internal/io_adapter.py (hunk 5 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/io_adapter.py:551]: resolved https://github.com/pytorch/pytorch/pull/113477#discussion_r1395014051

File: test/onnx/onnx_test_common.py (hunk 4 above)
Comment (thiagocrepaldi) [reply test/onnx/onnx_test_common.py:466]: > I didn't do this because it felt non-trivial. onnx outputs are more organized. 
> It can only be tensors, or (nested) list of tensors. Whereas for torch, the output can be anything. So mapping back from onnx to torch is considerably harder (how to insert back the Nones at the correct location where we discarded? how to construct back the arbitrary torch model output classes/datastructures that we flattened?).

That is a good point. My PR doesn't include a test with a very complex output that need to be rebuilt. This PR only has a test which requires buffers to be prepended to the original output.

I am wondering whether `pytree` could do that for us since it records the spec of the original data?

> I see that the goal was rather to eliminate the extra mutated buffers. I feel finding the way to include those outputs, when they exist because we are exporting ExportProgram, in the `ref_outputs` and have them properly accuracy checked might be more desirable.

We already do that too :) `adapt_onnx_outputs_to_torch` is not strictly necessary to support mutated buffers, but because you had proposed in another [review](https://github.com/pytorch/pytorch/pull/109650#discussion_r1331940092) I took this opportunity to implement here and improve the UX

Maybe we can split this PR with the essentials for supporting mutated buffers and move the `adapt_onnx_outputs_to_torch` into a new change we might need to dig deeper

File: test/onnx/onnx_test_common.py [hunk 6]
```diff
@@ -442,14 +442,24 @@ def _compare_pytorch_onnx_with_ort(
     ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(
         ref_model(*ref_input_args, **ref_input_kwargs)
     )
-
     ort_outputs = onnx_program(*input_args, **input_kwargs)
 
-    if len(ref_outputs) != len(ort_outputs):
+    # When model is a torch.export.ExportedProgram, the number of outputs in the ONNX model can be greater
```
Comment (thiagocrepaldi) [inline test/onnx/onnx_test_common.py:447]: https://github.com/pytorch/pytorch/pull/113686 will improve this

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 7]
```diff
@@ -942,6 +942,32 @@ def forward(self, x):
             loaded_exported_program, (x,), skip_dynamic_shapes_check=True
         )
 
+    @pytorch_test_common.xfail_if_model_type_is_not_exportedprogram(
+        "Unsupported FX nodes: {'call_function': ['aten.add_.Tensor']}. "
```
Comment (titaiwangms) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:946]: @BowenBao should Functionalization take care of this?

Comment (titaiwangms) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:946]: Looks like discrepency between our Functionalization and the one inside torch.export.export?

File: test/onnx/onnx_test_common.py [hunk 8]
```diff
@@ -436,16 +436,18 @@ def _compare_pytorch_onnx_with_ort(
...
-    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(
-        ref_model(*ref_input_args, **ref_input_kwargs)
-    )
-
+    # ONNXProgram holds a reference (not copy) to the original ref_model, including its state_dict.
+    # Thus, ONNXProgram() must run before ref_model() to prevent ref_model.forward() from changing the state_dict.
+    # Otherwise, the ref_model can change buffers on state_dict which would be used by ONNXProgram.__call__()
     ort_outputs = onnx_program(*input_args, **input_kwargs)
+    ref_outputs = ref_model(*ref_input_args, **ref_input_kwargs)
+    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_outputs)
```
Comment (titaiwangms) [inline test/onnx/onnx_test_common.py:444]: Why wasn't this a requirement? Could you elaborate more?

Comment (thiagocrepaldi) [reply test/onnx/onnx_test_common.py:444]: there 2 changes here.

one is just aesthetic, not thing really changed. 
Before we had `    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_model(*ref_input_args,**ref_input_kwargs))` and now we have 
```python
    ref_outputs = ref_model(*ref_input_args, **ref_input_kwargs)
    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_outputs)
```

The second change is swapping the order between `ref_outputs` and `ort_outputs` which is explained by the new comment:
```python
    # ONNXProgram holds a reference (not copy) to the original ref_model, including its state_dict.
    # Thus, ONNXProgram() must run before ref_model() to prevent ref_model.forward() from changing the state_dict.
    # Otherwise, the ref_model can change buffers on state_dict which would be used by ONNXProgram.__call__()
```

Comment (titaiwangms) [reply test/onnx/onnx_test_common.py:444]: My question was why we didn't need the second change? Because having the need of the second change doesn't feel very intuitive.

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py (hunk 7 above)
Comment (titaiwangms) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:946]: Make sure we have an issue tracking this.

**Review by titaiwangms — APPROVED** [review]
Please make a NOTE saying mutated buffers are no longer static after torch.export.export, and test if non-registered buffer is also dynamic. Thanks!

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py (hunk 7 above)
Comment (thiagocrepaldi) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:946]: https://github.com/pytorch/pytorch/issues/114406

File: test/onnx/onnx_test_common.py (hunk 8 above)
Comment (thiagocrepaldi) [reply test/onnx/onnx_test_common.py:444]: Offline discussion. When the input model is a `torch.nn.Module`, the FX tracer cannot handle buffer mutation, so the order in which ORT model and Torch model were never an issue. When the input model is `torch.export.ExportedProgram`, we internally use AOTAutograd which lifts mutated buffers as model inputs/outputs, so we need to execute ORT (with the pytorch model state) before torch to make sure both models are on the same iteration (forward step).

File: test/onnx/onnx_test_common.py [hunk 9]
```diff
@@ -436,16 +436,18 @@ def _compare_pytorch_onnx_with_ort(
         ref_input_args = input_args
         ref_input_kwargs = input_kwargs
 
-    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(
-        ref_model(*ref_input_args, **ref_input_kwargs)
-    )
-
+    # ONNXProgram holds a reference (not copy) to the original ref_model, including its state_dict.
+    # Thus, ONNXProgram() must run before ref_model() to prevent ref_model.forward() from changing the state_dict.
+    # Otherwise, the ref_model can change buffers on state_dict which would be used by ONNXProgram.__call__()
```
Comment (BowenBao) [inline test/onnx/onnx_test_common.py:441]: I recall there was a PR that will make ONNXProgram.__call__() update original pytorch model lol. This problem will arise again if that's added.

Btw there is a flag `has_mutation` up there that triggers model cloning. Might be helpful.

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py (hunk 7 above)
Comment (BowenBao) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:946]: yea because dynamo_export was not lifting buffers, there was another issue you can link it there too.

**Review by BowenBao — APPROVED** [review]

**Comment by huydhn** [comment]
It looks like the revert doesn't help, I will reland the change.  Sorry for the churn.

File: test/onnx/onnx_test_common.py (hunk 9 above)
Comment (thiagocrepaldi) [reply test/onnx/onnx_test_common.py:441]: that pr was merged with this one.
